File: 0387-first-unique-character-in-a-string/0387-first-unique-character-in-a-string.cpp
class Solution {
public:
    int firstUniqChar(string s) {
        unordered_map< char, pair< int,int > > map;
        
        for(int i = 0;i<s.length();i++){
            if (map.find(s[i]) == map.end()) {
                map[s[i]] = {1, i};
            } else {
                map[s[i]].first++;
            }
        }
        int ans = INT_MAX ;
        for(auto itr : map) 
            if(itr.second.first==1) 
                ans = min(ans, itr.second.second);
        
        return ans == INT_MAX ? -1 : ans ;
    }
};

File: 0387-first-unique-character-in-a-string/0387-first-unique-character-in-a-string.rs
use std::collections::HashMap;
impl Solution {
    pub fn first_uniq_char(s: String) -> i32 {
        let mut map : HashMap<char,(i32,usize)> = HashMap :: new();
        
        for(i ,ch ) in s.chars().enumerate() {
            if !map.contains_key(&ch){
                map.insert(ch,(1,i));
            }
            else{
                if let Some(entry) = map.get_mut(&ch){
                    entry.0 += 1;
                }
            }
        }
        let mut ans = std :: usize :: MAX;
        for(_,&(freq,idx)) in &map {
            if freq == 1 {
                ans = std::cmp::min(ans,idx);
            }
        }
        
        if ans == std::usize::MAX {
            -1
        }
        else {
            ans as i32
        }
    }
}